mini-projects/01-device-driver/laptop-interface/src/serial.rs

```diff
@@ -0,0 +1,159 @@
+        let name = sp_get_port_name(port);
+        let name_str = CStr::from_ptr(name).to_string_lossy();
+
+        // Check if the port name contains "usb"
+        if name_str.to_lowercase().contains("usb") {
+            ports.push(name_str.to_string());
+        }
+
+        i += 1;
+    }
+
+    sp_free_port_list(port_list); // Free the port list
+    if ports.is_empty() {
+        return Err("No USB ports found".to_string());
+    }
+
+    Ok(ports)
+}
+
+/// Initialize the serial connection to the specified port
+pub unsafe fn init_serial(port_name: &str) -> Result<*mut SpPort, String> {
+    let port_name_cstring = CString::new(port_name).unwrap();
+    let mut port_handle: *mut SpPort = ptr::null_mut();
+
+    let res = sp_get_port_by_name(port_name_cstring.as_ptr(), &mut port_handle);
+    if res != SpReturn::SP_OK {
+        return Err("Failed to get port by name".to_string());
+    }
+
+    let res = sp_open(port_handle, SP_MODE_READ_WRITE);
+    if res != SpReturn::SP_OK {
+        return Err("Failed to open port".to_string());
+    }
+
+    sp_set_baudrate(port_handle, 115200);
+    sp_set_bits(port_handle, 8);
+    sp_set_parity(port_handle, SpParity::SP_PARITY_NONE);
+    sp_set_stopbits(port_handle, 1);
+    sp_set_flowcontrol(port_handle, SpFlowcontrol::SP_FLOWCONTROL_NONE);
+
+    // Write "stop controller" to the port to make sure it's in the right state
+    let stop_controller = b"stop controller\n";
+    match blocking_write(port_handle, stop_controller) {
+        Ok(_) => println!(
+            "Wrote {} to the port {}",
+            std::str::from_utf8(stop_controller).unwrap(),
+            port_name
+        ),
+        Err(e) => eprintln!(
+            "Failed to write {} to {}: {}",
+            std::str::from_utf8(stop_controller).unwrap(),
+            port_name,
+            e
+        ),
+    }
+
+    thread::sleep(std::time::Duration::from_millis(100));
+
+    // Write "reset" to the port
+    let reset_controller = b"reset\n";
+    match blocking_write(port_handle, reset_controller) {
+        Ok(_) => println!(
+            "Wrote {} to the port {}",
+            std::str::from_utf8(reset_controller).unwrap(),
+            port_name
+        ),
+        Err(e) => eprintln!(
+            "Failed to write {} to {}: {}",
+            std::str::from_utf8(reset_controller).unwrap(),
+            port_name,
+            e
+        ),
+    }
+
+    thread::sleep(std::time::Duration::from_millis(100));
+
+    // Write "init controller" to the port
+    let init_controller = b"init controller\n";
+    match blocking_write(port_handle, init_controller) {
+        Ok(_) => println!(
+            "Wrote {} to the port {}",
+            std::str::from_utf8(init_controller).unwrap(),
+            port_name
+        ),
+        Err(e) => eprintln!(
+            "Failed to write {} to {}: {}",
+            std::str::from_utf8(init_controller).unwrap(),
+            port_name,
+            e
+        ),
+    }
```

Comment:
nit: this definitely could have been split into its own "write_str_to_serial" function so that you don't have to repeatedly do this error checking